app/org/sagebionetworks/bridge/services/backfill/UploadValidationByRecordIdBackfill.java

```diff
@@ -0,0 +1,90 @@
+package org.sagebionetworks.bridge.services.backfill;
+
+import com.amazonaws.services.s3.model.S3Object;
+import com.google.common.base.Charsets;
+import com.google.common.io.CharStreams;
+import org.sagebionetworks.bridge.config.BridgeConfigFactory;
+import org.sagebionetworks.bridge.dao.HealthDataDao;
+import org.sagebionetworks.bridge.models.backfill.BackfillTask;
+import org.sagebionetworks.bridge.models.healthdata.HealthDataRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * <p>
+ * Re-drives upload validation for a list of record IDs. These record IDs live in a file called
+ * upload-validation-backfill-recordIds in an S3 bucket called org-sagebridge-backfill-prod (replace prod) with your
+ * env name. The record IDs are then queried against the healthDataDao to get the upload ID.
+ * </p>
+ * <p>
+ * File format is one record ID per line.
+ * </p>
+ */
+@Component("uploadValidationByRecordIdBackfill")
+public class UploadValidationByRecordIdBackfill extends UploadValidationBackfill {
+    private static final Logger logger = LoggerFactory.getLogger(UploadValidationByRecordIdBackfill.class);
+
+    private static final String RECORD_ID_BUCKET = "org-sagebridge-backfill-" + BridgeConfigFactory.getConfig()
+            .getEnvironment().name().toLowerCase();
+    private static final String RECORD_ID_FILENAME = "upload-validation-backfill-recordIds";
+
+    private HealthDataDao healthDataDao;
+
+    /** Health data DAO, used to get the upload ID from the record ID. */
+    @Autowired
+    public void setHealthDataDao(HealthDataDao healthDataDao) {
+        this.healthDataDao = healthDataDao;
+    }
+
+    /** @{inheritDoc} */
+    @Override
+    protected List<String> getUploadIdList(BackfillTask task, BackfillCallback callback) throws IOException {
+        // get file of list of record IDs
+        S3Object recordIdFile = getS3Client().getObject(RECORD_ID_BUCKET, RECORD_ID_FILENAME);
```

Comment:
L51-56 duplicate the sibling class UploadValidationByUploadIdBackfill L35-39.  How about moving the code to the parent class to a new method that takes a bucket name and file name and returns a list of ids?  The two subclass can both use the same method.